From 7681063bcc21b7826dcaf98b8a255e2f058ca4cb Mon Sep 17 00:00:00 2001
From: Roy Pledge <roy.pledge@nxp.com>
Date: Fri, 26 Oct 2018 10:49:13 -0400
Subject: [PATCH 395/741] vfio/fsl-mc: Add support for shareable portals

commit a58c7148c2f97e508ac464dcb1cb4404c4297ca2 from
http://source.codeaurora.org/external/qoriq/qoriq-components/linux

Add support for mapping QBMan portals as shareable. This is
the mode used in QBMan version 5 and greater

Signed-off-by: Roy Pledge <roy.pledge@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/vfio/fsl-mc/vfio_fsl_mc.c         | 13 ++++++++++---
 drivers/vfio/fsl-mc/vfio_fsl_mc_private.h |  2 ++
 2 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/drivers/vfio/fsl-mc/vfio_fsl_mc.c b/drivers/vfio/fsl-mc/vfio_fsl_mc.c
index 9d134529f447..e343c6013429 100644
--- a/drivers/vfio/fsl-mc/vfio_fsl_mc.c
+++ b/drivers/vfio/fsl-mc/vfio_fsl_mc.c
@@ -59,6 +59,10 @@ static int vfio_fsl_mc_regions_init(struct vfio_fsl_mc_device *vdev)
 		if (mc_dev->regions[i].flags & IORESOURCE_CACHEABLE)
 			vdev->regions[i].type |=
 					VFIO_FSL_MC_REGION_TYPE_CACHEABLE;
+		if (mc_dev->regions[i].flags & IORESOURCE_MEM)
+			vdev->regions[i].type |=
+					VFIO_FSL_MC_REGION_TYPE_SHAREABLE;
+
 		vdev->regions[i].flags = VFIO_REGION_INFO_FLAG_MMAP;
 		vdev->regions[i].flags |= VFIO_REGION_INFO_FLAG_READ;
 		if (!(mc_dev->regions[i].flags & IORESOURCE_READONLY))
@@ -440,9 +444,12 @@ static int vfio_fsl_mc_mmap_mmio(struct vfio_fsl_mc_region region,
 	 * cache inhibited area of the portal to avoid coherency issues
 	 * if a user migrates to another core.
 	 */
-	if (region.type & VFIO_FSL_MC_REGION_TYPE_CACHEABLE)
-		vma->vm_page_prot = pgprot_cached_ns(vma->vm_page_prot);
-	else
+	if (region.type & VFIO_FSL_MC_REGION_TYPE_CACHEABLE) {
+		if (region.type & VFIO_FSL_MC_REGION_TYPE_SHAREABLE)
+			vma->vm_page_prot = pgprot_cached(vma->vm_page_prot);
+		else
+			vma->vm_page_prot = pgprot_cached_ns(vma->vm_page_prot);
+	} else
 		vma->vm_page_prot = pgprot_noncached(vma->vm_page_prot);
 
 	vma->vm_pgoff = (region.addr >> PAGE_SHIFT) + pgoff;
diff --git a/drivers/vfio/fsl-mc/vfio_fsl_mc_private.h b/drivers/vfio/fsl-mc/vfio_fsl_mc_private.h
index 34e757546e59..ead81cb72d2f 100644
--- a/drivers/vfio/fsl-mc/vfio_fsl_mc_private.h
+++ b/drivers/vfio/fsl-mc/vfio_fsl_mc_private.h
@@ -32,6 +32,8 @@ struct vfio_fsl_mc_region {
 	u32			flags;
 #define VFIO_FSL_MC_REGION_TYPE_MMIO  1
 #define VFIO_FSL_MC_REGION_TYPE_CACHEABLE  2
+#define VFIO_FSL_MC_REGION_TYPE_SHAREABLE  4
+
 	u32			type;
 	u64			addr;
 	resource_size_t		size;
-- 
2.17.1

